clients/drcachesim/tools/filter/record_filter.cpp

```diff
@@ -128,8 +128,11 @@ bool
 record_filter_t::parallel_shard_exit(void *shard_data)
 {
     per_shard_t *per_shard = reinterpret_cast<per_shard_t *>(shard_data);
-    input_entry_count_ += per_shard->input_entry_count;
-    output_entry_count_ += per_shard->output_entry_count;
+    {
+        std::lock_guard<std::mutex> guard(shard_exit_mutex_);
```

Comment:
Aggregation should be done at the end by a single thread, avoiding any need for synchronization points during parallel operation, which can ruin performance with no benefit.  All of our existing tools operate this way: they do not delete per-shard data in shard_exit, keeping it for aggregation in print_results or wherever.

Reply:
Synchronization in parallel_shard_exit may not be that bad. Since this lock is acquired only a few times (#shard or #threads-in-trace).

Delaying freeing of per-shard data and performing additional book-keeping for remembering the per_shard ptrs doesn't seem ideal either.

Reply:
Though remembering per_shard ptrs has the benefit of also allowing retrieving per-shard outputs.

Reply:
Done.